[#39] 이미지 API 추가

gombasan opened the pull request:
# 작업사항
- #39 

### 작업 내용
[feat : 이미지 업로드 기능 추가](https://github.com/Clover-Habiters/backend/commit/8a7126c6c34af1e9e8fac8f51817d8cf680838ca)  
- NCP 내에 ObjectStorage 로 이미지를 저장한 후에 이미지의 url 을 반환하도록 설정하였습니다.
- 이미지를 임시 경로인 tmp 파일에 저장해 둔 후에 스토리지 서버에 업로드 후 삭제합니다.

[feat(Test) : 이미지 업로드 테스트 추가](https://github.com/Clover-Habiters/backend/commit/d671f2b6eb459a6bb4905817426ba5b43d01939e)  
- Mock 객체를 사용하여 Url 을 정상적으로 반환 하는지 확인하는 테스트코드입니다.
- 만약 이미지 파일이 null 로 오게 된다면, IllegalArgumentException 예외를 발생시킵니다.
- RestDocs 로 문서화 내용을 추가하였습니다.

[feat : 썸네일 내용 추가](https://github.com/Clover-Habiters/backend/commit/636df3cb047a9af20543a7ba2bb4f35aad719da3)  
- 게시글을 조회 할 경우 썸네일 이미지의 url 정보를 추가하였습니다.
- 게시글을 등록 할 경우 썸네일 이미지의 url 을 같이 요청받아 등록하도록 하였습니다.
- 썸네일 이미지 url 은 옵션값으로 필수값은 아닙니다.

[feat(Test) : 썸네일 테스트 코드 추가](https://github.com/Clover-Habiters/backend/commit/e4645d2c38f68f1dd4c43fce68d4d2b21397848a) 
- 게시글 등록, 리스트 조회 시 썸네일 이미지가 정상적으로 등록,조회 테스트 코드를 추가하였습니다.
- 썸네일의 restDocs 문서화 내용을 추가하였습니다.
- PostProvider 에서 PostRequest 를 생성 할 경우 기본 썸네일 값을 추가하여 생성하도록 변경하였습니다.

[docs : 이미지 api 문서화 내용 추가](https://github.com/Clover-Habiters/backend/commit/e4246fb8d508319ed859ee8bb16073a8c2713565)  
- 이미지 api.adoc 파일 추가하였습니다.
- index 내에 링크를 추가하였습니다.

### 특이 사항
- 비동기 처리는 아직 시도해보지 못했습니다...!🥲🥲🥲
- 기존의 프로필 이미지 등록 코드를 재활용 하여 기능 추가하였습니다!!!
- 다운로드는 버킷에서 자유롭게 다운로드가 가능하여, 따로 api 를 만들지는 않았어요!
p.s. 아키텍쳐 아주 기가맥힙니다. 볼떄마다 감탄중이에여 ㅎㅎㅎ

JoosungKwon committed:
Update README.md

JoosungKwon committed:
README.md에 Architecture 이미지 추가

JoosungKwon committed:
Update README.md

gombasan committed:
feat : 이미지 업로드 기능 추가

- NCP 내에 ObjectStorage 로 이미지를 저장한 후에 이미지의 url 을 반환하도록 설정하였습니다.
- 이미지를 임시 경로인 tmp 파일에 저장해 둔 후에 스토리지 서버에 업로드 후 삭제합니다.

gombasan committed:
feat : 이미지 업로드 기능 추가

- NCP 내에 ObjectStorage 로 이미지를 저장한 후에 이미지의 url 을 반환하도록 설정하였습니다.
- 이미지를 임시 경로인 tmp 파일에 저장해 둔 후에 스토리지 서버에 업로드 후 삭제합니다.

gombasan committed:
feat(Test) : 이미지 업로드 테스트 추가

- Mock 객체를 사용하여 Url 을 정상적으로 반환 하는지 확인하는 테스트코드입니다.
- 만약 이미지 파일이 null 로 오게 된다면, IllegalArgumentException 예외를 발생시킵니다.
- RestDocs 로 문서화 내용을 추가하였습니다.

gombasan committed:
feat(Test) : 이미지 업로드 테스트 추가

- Mock 객체를 사용하여 Url 을 정상적으로 반환 하는지 확인하는 테스트코드입니다.
- 만약 이미지 파일이 null 로 오게 된다면, IllegalArgumentException 예외를 발생시킵니다.
- RestDocs 로 문서화 내용을 추가하였습니다.

gombasan committed:
docs : 이미지 api 문서화 내용 추가

- 이미지 api.adoc 파일 추가하였습니다.
- index 내에 링크를 추가하였습니다.

gombasan committed:
docs : 이미지 api 문서화 내용 추가

- 이미지 api.adoc 파일 추가하였습니다.
- index 내에 링크를 추가하였습니다.

gombasan committed:
feat : 썸네일 내용 추가

- 게시글을 조회 할 경우 썸네일 이미지의 url 정보를 추가하였습니다.
- 게시글을 등록 할 경우 썸네일 이미지의 url 을 같이 요청받아 등록하도록 하였습니다.
- 썸네일 이미지 url 은 옵션값으로 필수값은 아닙니다.

gombasan committed:
feat : 썸네일 내용 추가

- 게시글을 조회 할 경우 썸네일 이미지의 url 정보를 추가하였습니다.
- 게시글을 등록 할 경우 썸네일 이미지의 url 을 같이 요청받아 등록하도록 하였습니다.
- 썸네일 이미지 url 은 옵션값으로 필수값은 아닙니다.

gombasan committed:
feat(Test) : 썸네일 테스트 코드 추가

- 게시글 등록, 리스트 조회 시 썸네일 이미지가 정상적으로 등록,조회 테스트 코드를 추가하였습니다.
- 썸네일의 restDocs 문서화 내용을 추가하였습니다.
- PostProvider 에서 PostRequest 를 생성 할 경우 기본 썸네일 값을 추가하여 생성하도록 변경하였습니다.

gombasan committed:
feat(Test) : 썸네일 테스트 코드 추가

- 게시글 등록, 리스트 조회 시 썸네일 이미지가 정상적으로 등록,조회 테스트 코드를 추가하였습니다.
- 썸네일의 restDocs 문서화 내용을 추가하였습니다.
- PostProvider 에서 PostRequest 를 생성 할 경우 기본 썸네일 값을 추가하여 생성하도록 변경하였습니다.

JoosungKwon commented:
@gombasan 오타나서 CI 실패했네요!

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/post/dto/PostRequest.java:
```diff
@@ -5,6 +5,7 @@
 public record PostRequest(
 	String title,
 	String content,
+	String thumbnail,
```
thumbnail url이 더 나을 것 같아요!

gombasan replied:
넵! 저두 생각해보니 url 이라는 내용이 추가되어야 할거같네요 ㅎㅎ
수정하겠습니다~!

JoosungKwon commented on src/main/java/com/clover/habbittracker/global/infra/storage/api/ObjectStorageController.java:
```diff
@@ -0,0 +1,29 @@
+import lombok.RequiredArgsConstructor;
+
+@RestController
+@RequestMapping("/image")
```
저는 url 기본을 복수형으로 두는 데 어떻게 생각하시나요? ex) `images`

gombasan replied:
앗....! 저두 url은 복수로 사용하는게 좋다고 생각합니다!
url을 크게 신경안쓰고 코드 추가를 했네요.. 수정하도록 하겠습니다!

JoosungKwon commented on src/main/java/com/clover/habbittracker/global/infra/storage/service/ObjectStorageService.java:
```diff
@@ -0,0 +1,40 @@
+public class ObjectStorageService {
+
+	private final AmazonS3 objectStorage;
+	private final String BUCKET_NAME = "post-images";
```
Buket Name은 NCP와 동일한 설정 값이니 `application.yml`에 설정하는 것은 어떤가요?

gombasan replied:
오호 그게 더 좋을거같아요! ㅎㅎ 수정하도록 하겠습니다!

JoosungKwon commented on src/main/java/com/clover/habbittracker/global/infra/storage/service/ObjectStorageService.java:
```diff
@@ -0,0 +1,40 @@
+	private final String BUCKET_NAME = "post-images";
+
+	public String imgSave(MultipartFile file) {
+		Optional.ofNullable(file).orElseThrow(IllegalArgumentException::new);
```
null 처리가 되는 건가요?

gombasan replied:
테스트 돌려봤을땐 null 파일이 왔을경우 예외를 발생시킵니다!! 잘 처리 되는거같습니다!

JoosungKwon replied:
아 그런가요? 신기하네욥.. if문으로 처리하는 거하고 어떤게 성능상 이점이 있을지 궁금하네요..🤔

JoosungKwon replied:
아 그런가요? 신기하네욥.. if문으로 처리하는 거하고 어떤게 성능상 이점이 있을지 궁금하네요..🤔

gombasan replied:
제가 알기로는 Optinal 로 감싸고 `null`  일 경우 `Optional.empty()` 로 넘어와서 NPE 가 터지지 않는 것을 알고있습니다!
성능상 이점은.... if 문으로 null을 바로 처리해주는게 더 좋지않을까 싶네용...

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/infra/storage/api/ObjectStorageControllerTest.java:
```diff
@@ -0,0 +1,50 @@
+
+	@BeforeEach
+	void setUp() {
+		MockitoAnnotations.openMocks(this);
```
`MockitoExtension`으로 설정할 수 있습니다!

gombasan replied:
Mockito 에 대해 좀 공부가 필요할거같습니다...🥲

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/infra/storage/api/ObjectStorageControllerTest.java:
```diff
@@ -0,0 +1,50 @@
+	@BeforeEach
+	void setUp() {
+		MockitoAnnotations.openMocks(this);
+		when(objectStorageService.imgSave(any(MultipartFile.class))).thenReturn("/sucess/url");
```
`success` 이부분 오타 났네요!

gombasan replied:
그렇네요!! 아니 이런거까지 오타검증을 해주다니... CI 굉장히 좋네요 ㅎㅎㅎ

JoosungKwon reviewed (commented):
고생 많으셨습니다!! 제가 중간에 README를 여러번 변경해서 커밋이 꼬였네요! 리베이스로 정상화 해두겠습니다~

JoosungKwon committed:
fix: 오타 수정

gombasan committed:
chore : Post 썸네일 변수명 변경

- 보다 정확한 정보를 사용하고자 `thumbnail` -> `thumbnailUrl` 로 변경하였습니다.
- DB 스키마도 같이 변경하였습니다.

gombasan committed:
chore : 이미지 업로드 url 수정 및 오타 수정

- `image` -> `images` 로 url 을 변경하였습니다.
- `sucess` -> `success` 로 오타를 수정하였습니다.

gombasan committed:
chore : 버킷 이름을 설정으로 변경

- 버킷 이름은 변수보다는 설정값으로 사용되므로 설정파일로 값을 수정하였습니다.

gombasan committed:
Merge remote-tracking branch 'origin/feat/39-add-image' into feat/39-add-image

# Conflicts:
#	README.md
#	src/main/java/com/clover/habbittracker/domain/post/dto/PostRequest.java
#	src/main/java/com/clover/habbittracker/domain/post/dto/PostResponse.java
#	src/main/java/com/clover/habbittracker/domain/post/entity/Post.java
#	src/main/java/com/clover/habbittracker/global/infra/storage/api/ObjectStorageController.java
#	src/main/java/com/clover/habbittracker/global/infra/storage/service/ObjectStorageService.java
#	src/main/resources/schema.sql
#	src/test/java/com/clover/habbittracker/domain/post/api/PostControllerTest.java
#	src/test/java/com/clover/habbittracker/domain/post/service/PostServiceTest.java
#	src/test/java/com/clover/habbittracker/global/infra/storage/api/ObjectStorageControllerTest.java
#	src/test/java/com/clover/habbittracker/global/infra/storage/service/ObjectStorageServiceTest.java
#	src/test/java/com/clover/habbittracker/util/PostProvider.java

JoosungKwon commented:
아,, 제가 꼬인거 하드 푸시로 날렸는데 다시 꼬였네여,, 일단 머지하시고 제가 다시 수정할게요,,,

gombasan commented:
> 아,, 제가 꼬인거 하드 푸시로 날렸는데 다시 꼬였네여,, 일단 머지하시고 제가 다시 수정할게요,,,

ㅋㅋㅋㅋㅋㅋㅋㅋ... 제가 확인을 안하고 마구잡이로 푸시해서 생기는 문제입니다... 껄껄껄....
잘 확인하도록 할게용
머지하겠습니다!
